File: backend/models/PropertyJoinRequest.js
const mongoose = require('mongoose');

const STATUSES = ['pending', 'approved', 'rejected'];

// A prospective tenant's self-submitted request to join a property, filled
// out on the public web form at GET /invite?id=<propertyId> (no sign-in
// required - that's the point, they may not have the app yet). The owner
// reviews it in-app and approves/rejects; approving creates a real Tenant
// record (see routes/properties.js) the same way adding one manually does,
// including a fresh join code the owner then shares back.
const propertyJoinRequestSchema = new mongoose.Schema(
  {
    property: { type: mongoose.Schema.Types.ObjectId, ref: 'Property', required: true, index: true },
    name: { type: String, required: true, trim: true },
    phone: { type: String, required: true, trim: true },
    altPhone: { type: String, trim: true },
    roomNumber: { type: String, trim: true },
    monthlyRent: { type: Number, default: null },
    securityDeposit: { type: Number, default: null },
    moveInDate: { type: Date, default: null },
    status: { type: String, enum: STATUSES, default: 'pending' },
    // Set when approved, so a re-fetch of an already-handled request can
    // still surface which Tenant record it became.
    tenant: { type: mongoose.Schema.Types.ObjectId, ref: 'Tenant', default: null },
  },
  { timestamps: true }
);

const PropertyJoinRequest = mongoose.model('PropertyJoinRequest', propertyJoinRequestSchema);
PropertyJoinRequest.STATUSES = STATUSES;

module.exports = PropertyJoinRequest;

File: lib/models/property_join_request.dart
/// A prospective tenant's self-submitted request to join a property,
/// filled out on the public web form the "Invite Tenant" QR/link opens
/// (see backend/public/invite.html) — no sign-in required on their end.
/// The owner reviews it here; approving turns it into a real [Tenant].
class PropertyJoinRequest {
  final String id;
  final String name;
  final String phone;
  final String? altPhone;
  final String? roomNumber;
  final num? monthlyRent;
  final num? securityDeposit;
  final DateTime? moveInDate;
  final DateTime createdAt;

  const PropertyJoinRequest({
    required this.id,
    required this.name,
    required this.phone,
    this.altPhone,
    this.roomNumber,
    this.monthlyRent,
    this.securityDeposit,
    this.moveInDate,
    required this.createdAt,
  });

  factory PropertyJoinRequest.fromJson(Map<String, dynamic> json) => PropertyJoinRequest(
        id: json['_id'] as String,
        name: json['name'] as String,
        phone: json['phone'] as String,
        altPhone: json['altPhone'] as String?,
        roomNumber: json['roomNumber'] as String?,
        monthlyRent: json['monthlyRent'] as num?,
        securityDeposit: json['securityDeposit'] as num?,
        moveInDate: json['moveInDate'] != null ? DateTime.parse(json['moveInDate'] as String) : null,
        createdAt: DateTime.parse(json['createdAt'] as String),
      );
}
